design/accepted/gc_plus/uncomitted-gc.md

```diff
@@ -0,0 +1,215 @@
+enables additional optimizations on the GC process (read further for details).
+To resolve indexing issues with existing repositories that have a flat filesystem, suggested to create these slices under a `data` prefix. 
+
+![Repository Structure](uncommitted-gc-repo-struct.png)
+
+### 2. Listing of lakeFS repository committed objects
+
+Similar to the way GC works today, use repository meta-ranges and ranges to read all committed objects in the repository. 
+
+### 3. Listing of lakeFS repository uncommitted objects
+
+Expose a new API in lakeFS which writes repository uncommitted objects information into a parquet file in a dedicated path
+in the repository namespace
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+Uncommitted GC must scan the bucket in order to find objects that are not referenced by lakeFS.
+To optimize this process, suggest the following changes:
+
+1. Store lakeFS data under `<namespace>/data/` path
+2. Divide the repository data path into time-and-size based slices.
+3. Slice name will be a time based, reverse sorted unique identifier.
+4. LakeFS will create a new slice on a timely basis (for example: hourly) or when it has written < MAX_SLICE_SIZE > objects to the slice.
+5. Each slice will be written by a single lakeFS instance in order to track slice size.
+6. The sorted slices will enable partial scans of the bucket when running the optimized GC.
+
+#### StageObject
+
+The StageObject operation will only be allowed on addresses outside the repository's storage namespace. This way, 
+objects added using this operation are never collected by GC.
+
+#### [Get/Link]PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
+2. The token will be valid for a specified amount of time and for a single use.
+3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
+4. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+>**Note:** These changes will also solves the following [issue](https://github.com/treeverse/lakeFS/issues/4438)
+
+#### Track copied objects in ref-store
+
+lakeFS will track copy operations of uncommitted objects and store them in the ref-store for a limited duration.
+GC will use this information as part of the uncommitted data to avoid a race between the GC job and rename operation.
```

Comment:
We are tracking copy operations but the purpose is to avoid a race in rename.
Consider the following:
1. GC starts reading branch staging area 
2. Rename operation create a copy of staged entry - which doesn't not get listed due to listing order
3. Rename operation deletes original entry before GC is able to scan it as part of the listing of staging area

As a result, the physical address will not be listed in uncommitted (or committed) and will be candidate for deletion.
Reading from the copy table - allows exempting copied objects from the delete list and avoid this race.

The copied entries lifespan can be minutes or can be removed as part of a branch operation (commit/reset/delete branch)

In the next GC iteration - these addresses will either be in committed, still in staging or deleted and will be handled accordingly